dns_server/src/crawler/mod.rs

```diff
@@ -0,0 +1,560 @@
+
+/// When the server drops the unreachable node address that was once reachable. This should take about a month.
+/// Such a long time is useful if the server itself has prolonged connectivity problems.
+const PURGE_REACHABLE_FAIL_COUNT: u32 = 35;
+
+const STORAGE_VERSION: u32 = 1;
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressState {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+/// Additional state of a potential node address
+struct AddressData {
+    /// Connection state
+    state: AddressState,
+
+    /// Last time when the connection state was updated
+    state_updated_at: tokio::time::Instant,
+
+    /// The number of failed connection attempts.
+    /// New connection attempts are made after a progressive backoff time.
+    /// Resets to 0 when an outbound connection to the address is successful.
+    fail_count: u32,
+
+    /// Whether this address was reachable at least once.
+    /// Addresses that were once reachable are stored in the DB.
+    was_reachable: bool,
+}
+
+#[derive(Clone)]
+pub struct CrawlerConfig {
+    pub add_node: Vec<String>,
+
+    pub network: [u8; 4],
+
+    pub p2p_port: u16,
+}
+
+pub struct Crawler<N: NetworkingService, S> {
+    /// Crawler config
+    config: CrawlerConfig,
+
+    /// Backend's ConnectivityHandle
+    conn: N::ConnectivityHandle,
+
+    /// Backend's SyncingMessagingHandle
+    sync: N::SyncingMessagingHandle,
+
+    /// Map of all known addresses (including currently unreachable)
+    addresses: BTreeMap<N::Address, AddressData>,
+
+    /// Map of all currently connected peers
+    peers: BTreeMap<N::PeerId, N::Address>,
+
+    /// Storage implementation
+    storage: S,
+
+    /// Channel used to manage the DNS server
+    command_tx: mpsc::UnboundedSender<ServerCommands>,
+}
+
+impl<N: NetworkingService + 'static, S: DnsServerStorage> Crawler<N, S>
+where
+    N::SyncingMessagingHandle: SyncingMessagingService<N>,
+    N::ConnectivityHandle: ConnectivityService<N>,
+    DnsServerError: From<<<N as NetworkingService>::Address as FromStr>::Err>,
+{
+    pub fn new(
+        config: CrawlerConfig,
+        conn: N::ConnectivityHandle,
+        sync: N::SyncingMessagingHandle,
+        storage: S,
+        command_tx: mpsc::UnboundedSender<ServerCommands>,
+    ) -> Result<Self, DnsServerError> {
+        // Load all persistent addresses
+        let tx = storage.transaction_ro()?;
+
+        let storage_version = tx.get_version()?;
+        if storage_version.is_some() && storage_version != Some(STORAGE_VERSION) {
+            return Err(DnsServerError::Other("Unexpected storage version"));
+        }
+
+        let mut addresses = BTreeMap::new();
+        for address in tx.get_addresses()?.iter().filter_map(|address| address.parse().ok()) {
+            Self::new_address(&mut addresses, address, true);
+        }
+        tx.close();
+
```

Comment:
Please move all the selected code into a private function and return `addresses` from that.

Reply:
This code can be reordered to separate concerns much better than it looks now. I see storage_version is used later... so my exact suggestion from above isn't directly possible, but this can still be improved. Just know that whenever I see `mut` and then usage the alarms go off in my head.

Reply:
Right. Updated. I hope it's OK now.

Reply:
Updated one more time.